[DOCS-8243] Add session replay recorder explanation (#24962)
DataDog/documentation: content/en/real_user_monitoring/session_replay/mobile/_index.md

{{< img src="real_user_monitoring/session_replay/mobile/mobile_replay.mp4" alt="An example of a Mobile Session Replay recording" video="true" style="width:60%;">}}

## How the Session Replay recorder works

The Session Replay recorder is built into the RUM Mobile SDK. Unlike web browsers, mobile apps don't use HTML or CSS. Instead, the recorder takes a "snapshot" of your app's screen by breaking it into simple rectangles called "wireframes." It then keeps track of changes by only updating the wireframes that have changed, making the process efficient and fast.

### Wireframe concept

A _wireframe_ is like a digital sticky note that marks a specific area of your app's screen—such as a button, image, or background. Each wireframe is a rectangle that helps the recorder keep track of what's on the screen.

**Examples of wireframes:**
- A text label becomes a "text" wireframe, defined by its position and size.

> - The app's background is a "geometry" wireframe—a rectangle that covers the whole screen.

Reviewer: ```suggestion
- The app's background is a "shape" wireframe—a rectangle that covers the whole screen.
```

The first wireframe dictates the viewport size, enabling the Session Replay player to properly represent the device's screen size and orientation (landscape / portrait).

| Iteration | 4 | 5-11 | 12-13 |
|-----------|---|---|---|
| Viewport | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-4.png" alt="An example of a 'geometry', 'image', and 'text' wireframe." style="width:100%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-5.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:100%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-6.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:100%;">}} |

Because wireframes are sorted in back-to-front order, it overdraws the existing portion of the frame, which is a desirable behavior, as it helps support several UI patterns (for instance, semi-transparent elements).

| Iteration | 14-25 | Final result |
|-----------|-------|--------------|

> | Viewport | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-7.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:50%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-final.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:50%;">}} |

Reviewer: ```suggestion
| Viewport | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-7.png" alt="An example of a 'shape' and 'image' wireframe." style="width:50%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/iteration-final.png" alt="An example of a 'shape' and 'image' wireframe." style="width:50%;">}} |
```

{{< img src="real_user_monitoring/session_replay/mobile/mobile_replay.mp4" alt="An example of a Mobile Session Replay recording" video="true" style="width:60%;">}}

## How the Session Replay recorder works

> The Session Replay recorder is part of the RUM Mobile SDK. While [Browser Session Replay][1] relies on snapshots of the browser's DOM and CSS, while mobile apps don't have HTML, DOM, or CSS, so the recorder instead turns the native view's hierarchies into a sequence of flat "wireframes" and takes incremental snapshots of the relevant wireframes.

Reviewer 2: ```suggestion
The Session Replay recorder is part of the RUM Mobile SDK. While [Browser Session Replay][1] relies on snapshots of the browser's DOM and CSS, mobile apps don't have HTML, DOM, or CSS, so the recorder instead turns the native view hierarchy into a sequence of flat "wireframes", and takes incremental snapshots of the relevant wireframes.
```

### Wireframe concept

> A _wireframe_ describes individual rectangular areas in the mobile app screen. Wireframes are constructed during the process of traversing the app's view-tree in a bottom-up (or generally "back-to-front") order and determining visible elements. It is an abstract type, which means it doesn't always correspond 1:1 to a native view, or live inside the views hierarchy.

Reviewer 2: ```suggestion
A _wireframe_ describes individual rectangular areas in the mobile app screen. It is an abstract type, which means it doesn't always correspond 1:1 to a native view, or live inside the views hierarchy.
```

Removed this as it's already part of the "Recording algorithm" section

- The application's background view is considered a "geometry" wireframe stretched across the entire app window. In most cases, it can be described as a basic rectangle of a single color.
- Any other container view in the app with an opaque background is a "geometry" wireframe.
- An image or icon displayed in the app is an "image" wireframe with a defined position and sometimes style (such as opacity or alpha).
- A map view hierarchy consisting of dozens of views (for rendering shapes and annotations) can be merged into a single "image" wireframe.

### Recording algorithm

Wireframes are constructed during the process of traverseing the app's view-tree in a bottom-up (or generally "back-to-front") order and determining visible elements. In the below example, the iOS app screen is built with 78 native _views_, but it can be broken into 19 wireframes:

{{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-1.png" alt="An example of how the iOS app screen contains 78 native views, but is made up of 19 wireframes." style="width:30%;">}}

> These wireframes can be sent to the player while **preserving their rendering order** (back-to-front) and **using absolute positioning** (in screen coordinates). There is no "tree structure", nor child-parent relationship between wireframes, therefore making the structure "flat".

Reviewer 2: ```suggestion
These wireframes are recorded while **preserving their rendering order** (back-to-front) and **using absolute positioning** (in screen coordinates). There is no "tree structure", nor child-parent relationship between wireframes, therefore making the structure "flat".
```

### Rendering algorithm

> To display a single replay frame, the player needs to iterate through all wireframes and render one another into the viewport using geometry information (x, y, width, height) from each wireframe. It must respect wireframes rendering order, so that succeeding ones overdraw existing portions of the viewport.

Reviewer 2: ```suggestion
To display the Replay of a single frame on Datadog, the Session Replay player iterates through all wireframes and renders one another into the viewport using geometry information (x, y, width, height) from each one. **It respects wireframes rendering order**, so that succeeding ones overdraw existing portions of the viewport.
```

> The above screen can be reconstructed in 19 passes of the renderer:

Reviewer 2: ```suggestion
For instance, the screenshot displayed above is be reconstructed in 19 passes:
```

| Iteration | 1 | 2 | 3 |
|-----------|---|---|---|
| Viewport | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-1.png" alt="An example of a 'geometry' wireframe." style="width:100%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-2.png" alt="An example of an 'image' wireframe." style="width:100%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-3.png" alt="An example of a 'text' wireframe." style="width:100%;">}} |

> | Description | Draw "geometry" wireframe into viewport (rectangle + background color). | Draw "image" wireframe (rectangle + image with given UUID fetched from the Datadog backend). | Draw "text" wireframe (rectangle + "Twin Lake" text + font color, family and size). |

Reviewer 2: ```suggestion
```

Let me know what you think but I feel we can get rid of the in-table descriptions, that I feel are a bit too low level

Author: I would be ok with that if the visuals say enough!

> The first wireframe could dictate the viewport size, enabling player size adjustment for given device + rotation (landscape / portrait).

Reviewer 2: ```suggestion
The first wireframe dictates the viewport size, enabling the Session Replay player to properly represent the device's screen size and orientation (landscape / portrait).
```

| Iteration | 4 | 5-9 | 10-12 |
|-----------|---|---|---|
| Viewport | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-4.png" alt="An example of a 'geometry', 'image, and 'text' wireframe." style="width:100%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-5-9-1.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:100%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-10-12.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:100%;">}} |

> | Description | Draw "image" and "text" wireframes into viewport.| Draw "image" and "text" wireframes into viewport. | Draw "image" and two "geometry" wireframes into viewport.  |

Reviewer 2: ```suggestion
```

> Because wireframes are sorted in back-to-front order, it overdraws existing portion of the frame. This is very much desirable. For semi-transparent elements, it can also do the job of proper blending "closer" elements with the "farer" ones.

Reviewer 2: ```suggestion
Because wireframes are sorted in back-to-front order, it overdraws existing portion of the frame, which is a desirable behavior as it help support several UI patterns (for instance semi-transparent elements).
```

| Iteration | 13-19 | Final result |
|-----------|-------|--------------|
| Viewport | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-13-19.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:50%;">}} | {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-iteration-final-result.png" alt="An example of a 'geometry' and 'image' wireframe." style="width:50%;">}} |

### Full and incremental snapshots

> The sequence of all visible wireframes can be considered a full snapshot record if it is referring to the web format. Sending it per frame in the replay is not optimal, which is why the recorder takes incremental snapshots.
> 
> To send incremental changes, the mobile SDK tags each wireframe with a unique identifier. Native views are objects, sot hey can be identified by reference.

Reviewer 2: ```suggestion
To tie Mobile Session Replay back to its [Browser counterpart][1], the sequence of all visible wireframes is considered a full snapshot. But to avoid re-recording the entire screen when only incremental changes occur in the UI, the mobile SDK tags each wireframe with a unique identifier, and sends incremental snapshots to Datadog. Since native views are objects, they can be identified by reference.
```

> {{< img src="real_user_monitoring/session_replay/mobile/how-it-works/how-it-works-wireframe-uuid.png" alt="An example of how each wireframe is given a `uuid` (from 1 to 19)." style="width:40%;">}}
> 
> Except `uuid`, a wireframe can define 3 classes of attributes:
> 
> - **Geometry**: Its `x`, `y`, `width`, and `height` using absolute screen coordinates.
> - **Appearance**: Information on its background color, font size, font color, and other styles.
> - **Content**: Depending on wireframe type, it can be a text from the label, reference to an image (such as an image's `uuid` or remote URL) or any other information essential for its rendering.
> 

Reviewer 2: ```suggestion
```

I would remove the screenshot as it's redundant, and the piece about our data model as its too detailed and doesn't really provide any value for the reader.
